test/Infrastructure/LeanCode.AuditLogs.Tests/AuditLogsFilterTests.cs

```diff
@@ -0,0 +1,37 @@
+using MassTransit;
+using NSubstitute;
+using Xunit;
+
+namespace LeanCode.AuditLogs.Tests;
+
+public sealed class AuditLogsFilterTests
+{
+    public const string ConsumerName = "Test.Consumer.Name";
+
+    [Fact]
+    public async void Extracts_changes_after_pipeline_execution()
+    {
+        var dbContext = new TestDbContext();
+        var publisher = Substitute.For<AuditLogsPublisher>();
```

Comment:
This will result in the "real" logic of `AuditLogsPublisher` being run. IIRC NSubstitute won't by default stub any method of a class that isn't virtual, thus all the calls to `publisher.ExtractAndPublishAsync` will run (and you don't want that here).

Adding the missing `virtual` will fix the problem (and will probably fix the problem with randomly failing tests).